features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,993 @@
+        _active_area = area;
+        _active_area_version = versions[area];
+    }
+
+    // In case we have two empty areas, arbitrarily assign 0 to the active one.
+    if ((area_state[0] == AREA_STATE_EMPTY) && (area_state[1] == AREA_STATE_EMPTY)) {
+        _active_area = 0;
+        ret = write_master_record(_active_area, 1, _free_space_offset);
+        MBED_ASSERT(ret == NVSTORE_SUCCESS);
+        _init_done = 1;
+        return NVSTORE_SUCCESS;
+    }
+
+    // In case we have two valid areas, choose the one having the higher version (or 0
+    // in case of wrap around). Erase the other one.
+    if ((area_state[0] == AREA_STATE_VALID) && (area_state[1] == AREA_STATE_VALID)) {
+        if ((versions[0] > versions[1]) || (!versions[0]))
+            _active_area = 0;
+        else
+            _active_area = 1;
+        _active_area_version = versions[_active_area];
+        os_ret = flash_erase_area(1 - _active_area);
+        MBED_ASSERT(!os_ret);
+    }
+
+    // Traverse area until reaching the empty space at the end or until reaching a faulty record
+    while (_free_space_offset < free_space_offset_of_area[_active_area]) {
+        ret = read_record(_active_area, _free_space_offset, 0, NULL,
+                          actual_len_bytes, 1, valid,
+                          key, flags, next_offset);
+        MBED_ASSERT(ret == NVSTORE_SUCCESS);
+
+        // In case we have a faulty record, this probably means that the system crashed when written.
+        // Perform a garbage collection, to make the the other area valid.
+        if (!valid) {
+            ret = garbage_collection(NO_KEY, 0, 0, NULL);
+            break;
+        }
+        if (flags & DELETE_ITEM_FLAG)
+            _offset_by_key[key] = 0;
+        else
+            _offset_by_key[key] = _free_space_offset | (_active_area << OFFS_BY_KEY_AREA_BIT_POS) |
+                                    (((flags & SET_ONCE_FLAG) != 0) << OFFS_BY_KEY_SET_ONCE_BIT_POS);
+        _free_space_offset = next_offset;
+    }
+
+    _init_done = 1;
+    return NVSTORE_SUCCESS;
+}
+
+int NVStore::deinit()
+{
+    if (_init_done) {
+        nvstore_int_flash_deinit();
+        delete[] _offset_by_key;
+    }
+
+    _init_attempts = 0;
+    _init_done = 0;
+
+    return NVSTORE_SUCCESS;
+}
+
+int NVStore::reset()
+{
+    uint8_t area;
+    int os_ret;
+
+    // Erase both areas, and reinitialize the module. This is totally not thread safe,
+    // as init doesn't take the case of re-initialization into account. It's OK, as this function
+    // should only be called in pre-production cases.
+    for (area = 0; area < NVSTORE_NUM_AREAS; area++) {
+        os_ret = flash_erase_area(area);
+        if (os_ret)
+            return NVSTORE_WRITE_ERROR;
+    }
+
+    deinit();
+    return init();
+}
+
+uint32_t NVStore::size()
+{
+    if (!_init_done) {
+        init();
+    }
+
+    return _size;
+}
+
+#ifdef NVSTORE_TESTING
+
+int NVStore::force_garbage_collection(void)
```

Comment:
Whys is this bound with NVSTORE_TESTING? Shouldn't we be doing garbage collection for targets smaller flash footprint/smaller NVStore as well?

Reply:
Please note that this is is not the actual garbage collection function, which is simply called garbage_collection, but the *force* garbage collection function, used for testing purposes.

Reply:
@davidsaada Why would garbage collection need to be forced durring collection? Is this the same testing as `mbed test` or something different?